### PR TITLE
migrate Accept, Content-Type, Link, MediaType, MediaRange to cats-parse

### PR DESCRIPTION
**Pull request by novakov-alexey**
migrate to cats-parse:
- Accept
- Content-Type
- Link
- MediaType
- MediaRange

Some tests fail due to few TODOs:
- Link header depends on Rfc3986Parser (will be done by someone else or in a separate PR)
- ~~QValue parser from #4013 needs to be merged first~~

Part of #3984

### CODE REVIEW

**Comment by novakov-alexey** [comment]
There is one test case failing, I can't figure out why:

```bash
[error]   x Do a round trip through the Accept header (102 ms)
[error]    Accept: text/csv; quoteChar="'"; rowDelimiter=";"; escapeChar="\\\\\\\\"; charset="UTF-8"; columnDelimiter=" " != Accept: text/csv; quoteChar="'"; rowDelimiter=";"; escapeChar="\\\\"; charset="UTF-8"; columnDelimiter=" " (MediaTypeSpec.scala:55)
[error] org.http4s.MediaTypeSpec.$anonfun$new$16(MediaTypeSpec.scala:55)
[error] Actual:   ..."\\\\[\\\\]"; ch...
[error] Expected: ..."\\\\[]"; ch...
```
org.http4s.util.Writer#quote escapes every "\\" with one more "\\". 
Failing test is parsing header from already parsed header, so that we get escapeChar="\\\\\\\\" instead of expected escapeChar="\\\\". @rossabaker do you have an idea why that may happen?

**Review by rossabaker — COMMENTED** [review]
It looks like the parboiled2 version is returning a decoded quoted string, while the cats-parse version is returning the raw quoted string.

File: core/src/main/scala/org/http4s/headers/Content-Type.scala [hunk 1]
```diff
@@ -26,7 +28,35 @@ object `Content-Type` extends HeaderKey.Internal[`Content-Type`] with HeaderKey.
...
-    HttpHeaderParser.CONTENT_TYPE(s)
+    parser.parseAll(s).leftMap(e => ParseFailure("Invalid Content-Type header", e.toString))
+
+  private[http4s] val parser: Parser1[`Content-Type`] = {
+    import cats.parse.Parser._
+
+    val EOI = char('\uFFFF') // TODO: take this parser from AdditionalRules.scala
+    val EOL = optWs *> EOI.rep // TODO: take this parser from AdditionalRules.scala
+
+    (MediaRange.parser ~ MediaType.mediaTypeExtension.rep <* EOL).map {
```
[the remark this answers is not in the document]
Comment (novakov-alexey) [reply core/src/main/scala/org/http4s/headers/Content-Type.scala:39]: No, I do not think so. I guess it is some rudiment to delete.

**Comment by novakov-alexey** [comment]
> It looks like the parboiled2 version is returning a decoded quoted string, while the cats-parse version is returning the raw quoted string.

We can try to unescape it only for MediaType parser.

File: core/src/main/scala/org/http4s/headers/Content-Type.scala (hunk 1 above)
[the remark this answers is not in the document]
Comment (novakov-alexey) [reply core/src/main/scala/org/http4s/headers/Content-Type.scala]: Removed

**Comment by rossabaker** [comment]
I don't think Scala 3.0.0-M1 ever ran on JDK >= 14.  We can ignore that error.  While we're in flux, it would be nice to turn off the fail fast on CI.  We're going to have a hard time getting any results here until we fix that.

**Review by rossabaker — APPROVED** [review]
I think I'd probably rather have the `quotedString` helper take care of unescaping, but I'd need to spend more time making sure they always get interpreted the same way.

Hard to say given the current state of CI on these port branches, but I think this unblocks us?  I'm :+1:, and it's not going to a prod branch yet, so there's time to clean up if we forgot something.

**Comment by novakov-alexey** [comment]
Yes, I agree that unquoting could be done nicely via separate parser.

**Pull request merged**